### PR TITLE
Allow .onLoad and other special functions regardless of name style

### LINKED ISSUE #500: `object_name_linter()` complains about ns-hooks (e.g. `.onLoad`)
Reprex:

```r
example_file <- tempfile()

readr::write_file(".onLoad <- function(libname, pkgname) {\n  # some action\n}\n",
                  path = example_file)

lintr::lint(filename = example_file, linters = lintr::object_name_linter())
```

R's [hooks for namespace events](https://rdrr.io/r/base/ns-hooks.html) are CamelCase and you can't change that AFAIK, so lintr complaining about

```
Variable and function name style should be snake_case.
```

is pointless.

### PR DESCRIPTION
**Pull request by AshesITR**
Also remove relevant # nolint in lintrs own .onLoad.

fixes #500

### CODE REVIEW

File: R/object_name_linters.R [hunk 1]
```diff
@@ -240,6 +240,7 @@ is_base_function <- function(x) {
   x %in% base_funs
 }
 
+# see ?".onLoad" and ?"Startup"
```
Comment (AshesITR) [inline R/object_name_linters.R:243]: Thanks, very good idea to note where they come from :-)

File: R/object_name_linters.R [hunk 2]
```diff
@@ -129,7 +129,8 @@ make_object_linter <- function(fun) {
     keep_indices <- which(
       !is_operator(names) &
         !is_known_generic(names) &
-        !is_base_function(names)
+        !is_base_function(names) &
+        !is_special_function(names)
```
Comment (MichaelChirico) [inline R/object_name_linters.R:133]: Is there a better name than `special`? maybe `special_hook`? "special" seems a bit generic. If no better option springs to mind, just go ahead.

**Review by MichaelChirico — DISMISSED** [review]

File: R/object_name_linters.R (hunk 2 above)
Comment (MichaelChirico) [reply R/object_name_linters.R:133]: (also for the NEWS)

Comment (AshesITR) [reply R/object_name_linters.R:133]: I thought if we ever need to add additional functions to this list for whatever reason, we might not want to blow up the conditional and instead just update the list of special functions with more comments on why they're special.

Comment (MichaelChirico) [reply R/object_name_linters.R:133]: OK that makes sense. I was going to suggest improving the NEWS but I think that's done now too.

**Comment by AshesITR** [comment]
@MichaelChirico can you re-approve (merging master wasn't clean)? Thanks!

**Review by MichaelChirico — APPROVED** [review]

File: R/object_name_linters.R [hunk 3]
```diff
@@ -239,6 +240,21 @@ is_base_function <- function(x) {
...
+  ".onLoad",
+  ".onAttach",
+  ".onUnload",
+  ".onDetach",
+  ".Last.lib",
+  ".First",
+  ".Last"
+)
+
+is_special_function <- function(x) {
```
Comment (MichaelChirico) [inline R/object_name_linters.R:254]: hmm I guess we need a test for this :sweat_smile: I'm seeing this on current master when trying to do #597:

```
R/zzz.R:185:1: style: Variable and function name style should be snake_case.
.onLoad <- function(libname, pkgname) {
^~~~~~~
```

Comment (MichaelChirico) [reply R/object_name_linters.R:254]: The issue is that `strip_names` removes the leading `.`. So actually we need to search for `onLoad` etc.

**Pull request merged**